### ISSUE TITLE
Generic attributes of bound generic protocols are revealed as "Unknown".

### ISSUE THREAD

**Issue by Ovsyanka83** [microsoft/pyright#5177]
Note that this is not a bug per se because I see similar issues in mypy. I have consulted many people about this already and none have been able to figure out the reasoning behind this.

**Describe the bug**
Generic attributes of bound generic protocols are revealed as "Unknown".

**Expected behavior**
The attributes should get revealed as generics or as the types they were bound to.

**Screenshots or Code**
```python
from typing import Any, Callable, Concatenate, Generic, ParamSpec, Protocol, TypeVar

T_val = TypeVar("T_val", bound=int | None)


class Proto(Protocol, Generic[T_val]):
    value: T_val


T = TypeVar("T")
Self = TypeVar("Self", bound=Proto)
P = ParamSpec("P")
Meth = Callable[Concatenate[Self, P], T]


def meth_decorator(meth: Meth[Self, P, T]) -> Meth[Self, P, T]:
    def wrapper(self: Self, *args: P.args, **kwargs: P.kwargs) -> T:
        reveal_type(self.value)
        return meth(self, *args, **kwargs)

    return wrapper
```

**VS Code extension or command-line**
Pylance v2023.5.40

**Comment by erictraut** [comment]
Pyright's behavior here is correct. 

You've defined the type variable `Self` to be bound to the type `Proto[Unknown]` (where `Unknown` is an implied `Any`). If you enable strict-mode checking (or simply enable the `reportMissingTypeArgument` rule), you'll see that pyright complains about the fact that you haven't supplied a type argument for `Proto` in the `bound` argument. I presume that your intent is for it to be bound to `Proto[Any]`.

Not surprisingly, mypy produces the same results (although it outputs `Any` rather than `Unknown` because it doesn't distinguish between implicit and explicit `Any`).

**Comment by erictraut** [comment]
It sounds like you want the ability to bind a type variable to another generic type, such as `Proto[T]`. That's not allowed in the Python type system today (PEP 484 expressly forbids it). This would require support for a feature called "higher-kinded types" (HKTs) which allow type parameters to be parameterized by other type parameters. This is a very advanced feature that doesn't appear in most type systems, and I think it's unlikely that it will be added to Python any time soon.
